src/bvh/bvh_impl.rs

```diff
@@ -4,455 +4,81 @@
-            (child_l_index, child_l_aabb, child_r_index, child_r_aabb)
-        };
-
-        // Construct the actual data structure and replace the dummy node.
-        assert!(!child_l_aabb.is_empty());
-        assert!(!child_r_aabb.is_empty());
-        nodes[node_index] = BvhNode::Node {
-            parent_index,
-            child_l_aabb,
-            child_l_index,
-            child_r_aabb,
-            child_r_index,
-        };
-
-        node_index
-    }
-
-    /// Traverses the [`Bvh`] recursively and returns all shapes whose [`Aabb`] is
-    /// intersected by the given [`Ray`].
-    ///
-    /// [`Aabb`]: ../aabb/struct.Aabb.html
-    /// [`Bvh`]: struct.Bvh.html
-    /// [`Ray`]: ../ray/struct.Ray.html
-    ///
-    pub fn traverse_recursive(
-        nodes: &[BvhNode<T, D>],
-        node_index: usize,
-        ray: &Ray<T, D>,
-        indices: &mut Vec<usize>,
-    ) where
-        T: PartialOrd + Zero + ClosedSub + ClosedMul + SimdPartialOrd,
-    {
-        match nodes[node_index] {
-            BvhNode::Node {
-                ref child_l_aabb,
-                child_l_index,
-                ref child_r_aabb,
-                child_r_index,
-                ..
-            } => {
-                if ray.intersects_aabb(child_l_aabb) {
-                    BvhNode::traverse_recursive(nodes, child_l_index, ray, indices);
-                }
-                if ray.intersects_aabb(child_r_aabb) {
-                    BvhNode::traverse_recursive(nodes, child_r_index, ray, indices);
-                }
-            }
-            BvhNode::Leaf { shape_index, .. } => {
-                indices.push(shape_index);
-            }
-        }
-    }
-}
+use super::{BvhNode, BvhNodeBuildArgs, ShapeIndex, Shapes};
 
 /// The [`Bvh`] data structure. Contains the list of [`BvhNode`]s.
 ///
 /// [`Bvh`]: struct.Bvh.html
 ///
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-pub struct Bvh<T: Scalar + Copy, const D: usize> {
+pub struct Bvh<T: BHValue, const D: usize> {
     /// The list of nodes of the [`Bvh`].
     ///
     /// [`Bvh`]: struct.Bvh.html
     ///
     pub nodes: Vec<BvhNode<T, D>>,
 }
 
-impl<T: Scalar + Copy, const D: usize> Bvh<T, D> {
+impl<T: BHValue, const D: usize> Bvh<T, D> {
     /// Creates a new [`Bvh`] from the `shapes` slice.
     ///
     /// [`Bvh`]: struct.Bvh.html
     ///
-    pub fn build<Shape: BHShape<T, D>>(shapes: &mut [Shape]) -> Bvh<T, D>
-    where
-        T: FromPrimitive + ToPrimitive + Float + ClosedSub + ClosedAdd + ClosedMul + SimdPartialOrd,
-    {
-        let indices = (0..shapes.len()).collect::<Vec<usize>>();
-        let expected_node_count = shapes.len() * 2;
+    pub fn build<Shape: BHShape<T, D>>(shapes: &mut [Shape]) -> Bvh<T, D> {
+        Self::build_with_executor(shapes, |left, right| {
+            left.build();
+            right.build();
+        })
+    }
```

Comment:
my reasoning for leaving build and adding build_par was to avoid a breaking change in behavior, but I don't have a strong preference. I can also see value in build just becoming parallel by default.

Reply:
If we do make build parallel by default then we should also make sure we are still exposing a buid_serial type of method because there are cases where you don't want the bvh to build in parallel all of the time

Reply:
Hmm that's a good point. Well, in that case we might as well leave it the way it is. :)